relay-server/src/actors/processor.rs

```diff
@@ -216,50 +137,92 @@ impl ProcessingGroup {
+        });
+        if !replay_items.is_empty() {
+            grouped_envelopes.push((
+                ProcessingGroup::Replay,
+                Envelope::from_parts(headers.clone(), replay_items),
+            ))
+        }
+
+        // Keep all the sessions together in one envelope.
+        let session_items = envelope.take_items_by(|item| {
+            item.ty() == &ItemType::Session || item.ty() == &ItemType::Sessions
+        });
+        if !session_items.is_empty() {
+            grouped_envelopes.push((
+                ProcessingGroup::Session,
+                Envelope::from_parts(headers.clone(), session_items),
+            ))
+        }
 
+        // Extract spans.
+        let span_items = envelope
+            .take_items_by(|item| item.ty() == &ItemType::Span || item.ty() == &ItemType::OtelSpan);
+        if !span_items.is_empty() {
+            grouped_envelopes.push((
+                ProcessingGroup::Span,
+                Envelope::from_parts(headers.clone(), span_items),
+            ))
+        }
+
+        // Extract all the items which require event into separate envelope.
+        let require_event_items = envelope.take_items_by(Item::requires_event);
         if !require_event_items.is_empty() {
             let group = if require_event_items
                 .iter()
-                .any(|item| item.ty() == &ItemType::Transaction)
+                .any(|item| item.ty() == &ItemType::Transaction || item.ty() == &ItemType::Profile)
             {
                 ProcessingGroup::Transaction
             } else {
                 ProcessingGroup::Error
             };
-            envelopes.push((
+            grouped_envelopes.push((
                 group,
                 Envelope::from_parts(headers.clone(), require_event_items),
             ))
         }
 
-        if !span_items.is_empty() {
-            envelopes.push((
-                ProcessingGroup::Span,
-                Envelope::from_parts(headers.clone(), span_items),
-            ))
-        }
-
-        if !session_items.is_empty() {
-            envelopes.push((
-                ProcessingGroup::Session,
-                Envelope::from_parts(headers.clone(), session_items),
-            ))
-        }
+        // Get the rest of the envelopes, one per item.
+        let envelopes = envelope.items_mut().map(|item| {
+            let headers = headers.clone();
+            let items: SmallVec<[Item; 3]> = smallvec![item.clone()];
+            let envelope = Envelope::from_parts(headers, items);
+            let item_type = item.ty();
+            let group = if item_type == &ItemType::CheckIn {
+                ProcessingGroup::CheckIn
+            } else if item_type == &ItemType::UserReport || item_type == &ItemType::ClientReport {
+                ProcessingGroup::UserReport
+            } else {
+                relay_log::error!("Unsupported ItemType in the envelope: {item_type}");
```

Comment:
```suggestion
                if item_type != &ItemType::Unknown { // Unknown item types are forwarded by design
                    relay_log::error!("Unsupported ItemType in the envelope: {item_type}");
                }
```

See https://github.com/getsentry/relay/blob/d17608ca5d0cb9b897d7c96e99b5a7cddb09ae82/relay-server/src/envelope.rs#L126-L129

Reply:
Better idea: We can have a separate enum variant for deliberately forwarded unknown item types, and one for ungroupble items. I.e.

```rust
enum ProcessingGroup {
    // ...
    ForwardUnknown, // This is where ItemType::Unknown goes. No processing is done to it.
    Ungrouped, // this is everything we failed to group -> should log an error.

Reply:
done, now skipping processing for all the  uknown item types